src/backend/core/warren/base_indicator.py

```diff
@@ -2,9 +2,59 @@
 from abc import ABC, abstractmethod
 from typing import List
 
+import pandas as pd
+import numpy as np
 from ralph.backends.http.lrs import LRSQuery
 
 
+def parse_raw_statements(raw_statements) -> pd.DataFrame:
+    """Parse LRS statements, explode the columns, and add a `date` column."""
+    flattened = pd.json_normalize(raw_statements)
+    # Disable chained assignment warning to make the transformation inplace
+    pd.options.mode.chained_assignment = None
+    # Transform timestamp column into a date with day (YYYY-MM-DD)
+    flattened.loc[:, "date"] = pd.to_datetime(flattened.loc[:, "timestamp"]).dt.date
+    df = add_unique_actor_id(flattened)
+    return df
+
+
+def add_unique_actor_id(statements: pd.DataFrame) -> pd.DataFrame:
+    """Depending on the xAPI statements, the actor can be identified in 4 different ways :
+    https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Data.md#details-4
+
+    This function creates a `unique_agent_id` column that uniquely identifies the agent.
+    """
+
+    columns_to_check = ["actor.account.name", "actor.account.homePage", "actor.mbox", "actor.openid"]
+
+    for column in columns_to_check:
+        if column not in statements.columns:
+            statements[column] = np.nan
+
+    def generate_unique_agent_id(row):
+        non_nan_columns = [col for col in columns_to_check if pd.notna(row[col])]
+        if len(non_nan_columns) == 0:
+            raise ValueError("No way of identifying the agent: all identification columns are NaN.")
+        elif len(non_nan_columns) > 1:
+            if "actor.account.name" in non_nan_columns and "actor.account.homePage" in non_nan_columns:
+                return f"{row['actor.account.name']}-{row['actor.account.homePage']}"
+            else:
+                raise ValueError(
+                    "Multiple ways of identification for agent: more than one identification column is filled.")
+        else:
+            return row[non_nan_columns[0]]
```

Comment:
Overall: I find unclear adding nan column, and then checking which columns are not nan. I would be more than happy to discuss it. Why not using set diff?

`non_nan_columns = COLUMNS_TO_CHECK - set(row.index)`

Reply:
This need to be tested, but some suggestions can be reused. I find it clearer to avoid nested `if`/`else`, it reduces cognitive complexity.

```suggestion
        non_nan_columns = COLUMNS_TO_CHECK - set(row.index)
        
        if not non_nan_columns:
                raise ValueError("No way of identifying the agent: all identification columns are NaN.")
        
        if len(non_nan_columns) == 1:
                return row[non_nan_columns[0]]

        if not {"actor.account.name", "actor.account.homePage"}.issubset(non_nan_columns):
                raise ValueError("Multiple ways of identification for agent: more than one identification column is filled.")

        return f"{row['actor.account.name']}-{row['actor.account.homePage']}"
```